### DISCUSSION TITLE
[Q/A][Java SDK (common)] Should the group version value be null and the empty string be treated the same?

### DISCUSSION THREAD

**Discussion by FoghostCn** [apache/dubbo#13915] [Question] [answered]
### Pre-check

- [X] I am sure that all the content I provide is in English.

### Apache Dubbo Component

Java SDK (apache/dubbo)

### Details

Should the group version value be null and the empty string be treated the same? in dubbo java group and version default to null if not set, but in golang it default to empty string

![image](https://github.com/apache/dubbo/assets/5020327/406715ab-2535-4b38-bccc-07fcd894a75d)

this may cause compatiable issue,some logic depends on this, such as 

https://github.com/apache/dubbo/blob/cf53c38179c13846cb760011758bd2035575fe4a/dubbo-common/src/main/java/org/apache/dubbo/common/ServiceKey.java#L48-L59

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by AlbumenJ** [comment, marked as answer]
They should be the same

**Comment by AlbumenJ** [comment]
Can you explain in which situation this will work as unexpected?

**Reply by FoghostCn** [reply]
this is because https://github.com/apache/dubbo/discussions/13914 

I found that there is an extra layer of invoker in this scenario. After troubleshooting, I found that it was caused by the equals method of `ServiceKey`.

<img width="1841" alt="image" src="https://github.com/apache/dubbo/assets/5020327/8713dc6e-f62e-4562-9359-d8e48afc0e19">


`!consumerProtocolServiceKey.isSameWith(matchedProtocolServiceKeys.get(0)` is true cause `shouldWrap` is true and add a `InstanceWrappedInvoker` to the invoker created by `protocol.refer`.

I don’t know if there are any side effects, but this issue(#13914) seems to have something to do with it.

https://github.com/apache/dubbo/blob/d5c484a64396bf763b1c057802a5656afe98eb9f/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java#L470-L501

**Reply by AlbumenJ** [reply]
https://github.com/apache/dubbo/discussions/13914 is related with injvm protocol. But this seems not related with it?

**Reply by FoghostCn** [reply]
Yeah, not the direct cause, actually I mean if there should use a InstanceWrappedInvoker in this scenario

**Reply by AlbumenJ** [reply]
What scenario?

**Reply by FoghostCn** [reply]
this https://github.com/apache/dubbo/discussions/13914

**Reply by AlbumenJ** [reply]
Actually, I don't know what is the relation about https://github.com/apache/dubbo/discussions/13914 and #13915. 😭

**Reply by FoghostCn** [reply]
OK, Since #13914 has been solved, this matter is no longer important.

**Reply by AlbumenJ** [reply]
Do you mean https://github.com/apache/dubbo/pull/13932?

**Reply by FoghostCn** [reply]
Actually #13932 and #13914 are the same，they all refer to `InstanceWrappedInvoker`

https://github.com/apache/dubbo/blob/cf53c38179c13846cb760011758bd2035575fe4a/dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java#L776-L794
 
`InstanceWrappedInvoker` is used when `org.apache.dubbo.common.ProtocolServiceKey#isSameWith` is `true`, in #13932 and #13914 `InstanceWrappedInvoker` is used because group version value null and empty string is not the same in current implement

now we fixed `InstanceWrappedInvoker` in #13932, whatever `InstanceWrappedInvoker` is used or not it will no problems
